<a href="https://colab.research.google.com/github/rgiuffre90/DS-Unit-1-Sprint-2-Statistics/blob/master/Robert_Giuffre_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [1]:
# YOUR WORK HERE

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-08-11 18:40:35--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.04s   

2020-08-11 18:40:36 (528 KB/s) - ‘student.zip’ saved [20478/20478]



In [2]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [5]:
import pandas as pd
import numpy as np
from scipy import stats

In [6]:
df = pd.read_csv('student-mat.csv', sep=';')

df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


### Null Hypothesis

Total absences and failures are independent

### Alternative Hypothesis

Total absences and failures are dependent

In [7]:
contingency = pd.crosstab(df['failures'], df['absences'])

contingency

absences  0   1   2   3   4   5   6   7   ...  26  28  30  38  40  54  56  75
failures                                  ...                                
0         89   3  58   6  47   4  25   6  ...   1   0   1   0   0   1   1   1
1         14   0   4   1   4   0   3   1  ...   0   1   0   1   1   0   0   0
2          5   0   1   0   2   0   2   0  ...   0   0   0   0   0   0   0   0
3          7   0   2   1   0   1   1   0  ...   0   0   0   0   0   0   0   0

[4 rows x 34 columns]

In [8]:
observed_with_margins = pd.crosstab(df['failures'], df['absences'], margins=True)

observed_with_margins

absences    0  1   2  3   4  5   6  7   8  ...  26  28  30  38  40  54  56  75  All
failures                                   ...                                     
0          89  3  58  6  47  4  25  6  17  ...   1   0   1   0   0   1   1   1  312
1          14  0   4  1   4  0   3  1   2  ...   0   1   0   1   1   0   0   0   50
2           5  0   1  0   2  0   2  0   2  ...   0   0   0   0   0   0   0   0   17
3           7  0   2  1   0  1   1  0   1  ...   0   0   0   0   0   0   0   0   16
All       115  3  65  8  53  5  31  7  22  ...   1   1   1   1   1   1   1   1  395

[5 rows x 35 columns]

In [9]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 122.18899006084003
p value: 0.05694057065252934
dof: 99
expected_frequencies: 
 [[9.08354430e+01 2.36962025e+00 5.13417722e+01 6.31898734e+00
  4.18632911e+01 3.94936709e+00 2.44860759e+01 5.52911392e+00
  1.73772152e+01 2.36962025e+00 1.34278481e+01 2.36962025e+00
  9.47848101e+00 2.36962025e+00 9.47848101e+00 2.36962025e+00
  5.52911392e+00 7.89873418e-01 3.94936709e+00 7.89873418e-01
  3.15949367e+00 7.89873418e-01 2.36962025e+00 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01 7.89873418e-01 7.89873418e-01
  7.89873418e-01 7.89873418e-01]
 [1.45569620e+01 3.79746835e-01 8.22784810e+00 1.01265823e+00
  6.70886076e+00 6.32911392e-01 3.92405063e+00 8.86075949e-01
  2.78481013e+00 3.79746835e-01 2.15189873e+00 3.79746835e-01
  1.51898734e+00 3.79746835e-01 1.51898734e+00 3.79746835e-01
  8.86075949e-01 1.26582278e-01 6.32911392e-01 1.26582278e-01
  5.06329114e-01 1.26582278e-01 3.79746835e-01 1.26582278e-01
  1.26582278e

### Conclusion

Based on a chi2 statistic of 122.18 and a p-value of .056, we **fail to reject** the null hypothesis that absences and failures are independent

### Null Hypothesis

The type of guardian a student has and their total study time are independent

### Alternative Hypothesis

The type of guardian a student has and their total study time are dependent

In [10]:
contingency = pd.crosstab(df['studytime'], df['guardian'])

contingency

guardian   father  mother  other
studytime                       
1              26      73      6
2              44     137     17
3              10      47      8
4              10      16      1

In [11]:
observed_with_margins = pd.crosstab(df['studytime'], df['guardian'], margins=True)

observed_with_margins

guardian   father  mother  other  All
studytime                            
1              26      73      6  105
2              44     137     17  198
3              10      47      8   65
4              10      16      1   27
All            90     273     32  395

In [12]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 7.515041507665898
p value: 0.2758269993780965
dof: 6
expected_frequencies: 
 [[ 23.92405063  72.56962025   8.50632911]
 [ 45.11392405 136.84556962  16.04050633]
 [ 14.81012658  44.92405063   5.26582278]
 [  6.15189873  18.66075949   2.18734177]]


### Conclusion

Based on a chi2 statistic of 7.515 and a p-value of .275 we **reject** the null hypothesis that the type of guardian a student has is independent of total study time.

### Null Hypothesis

The decision to participate in activities is independent on travel time

### Alternative Hypothesis

The decision to participate in activites is dependent on travel time

In [14]:
contingency = pd.crosstab(df['activities'], df['traveltime'])

contingency

traveltime    1   2   3  4
activities                
no          126  52  12  4
yes         131  55  11  4

In [17]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 0.10084771353217425
p value: 0.9917355011344893
dof: 3
expected_frequencies: 
 [[126.22278481  52.55189873  11.29620253   3.92911392]
 [130.77721519  54.44810127  11.70379747   4.07088608]]


### Conclusion

Based on a chi2 value of .1008 and a p-value of .99 I **reject** the null hypothesis that the decision to participate in activites is independent of travel time.

### Null Hypothesis

The total study time of a student is independent of their sex


### Alternative Hypothesis

The total study time of a student is dependent of their sex

In [24]:
contingency = pd.crosstab(df['studytime'], df['sex'])

contingency

sex          F   M
studytime         
1           27  78
2          113  85
3           51  14
4           17  10

In [25]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 50.634037705651764
p value: 5.854045926336989e-11
dof: 3
expected_frequencies: 
 [[ 55.29113924  49.70886076]
 [104.26329114  93.73670886]
 [ 34.2278481   30.7721519 ]
 [ 14.21772152  12.78227848]]


### Conclusion

Based on a chi2 value of 50.63 and a p-value of .00000000000585 we **fail to reject** the null hypothesis that the total study time of a student is dependent of their sex

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# YOUR WORK HERE

In [60]:
row_totals = df['studytime'].value_counts().values

row_totals

array([198, 105,  65,  27])

In [61]:
column_totals = df['sex'].value_counts().sort_index().values

column_totals

array([208, 187])

In [66]:
sample_size = df.shape[0]

sample_size

395

In [67]:
expected = np.array([[0,0],
                    [0,0],
                    [0,0],
                    [0,0]])

for row_index, row in enumerate(row_totals):
  for column_index, col in enumerate(column_totals):
    expected_value = (row*col)/sample_size
    expected[row_index][column_index] = expected_value

    expected

In [68]:
observed = contingency.values

In [69]:
observed = ((observed - expected)**2 / expected).sum()

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

*it estimates some fact about the population. For example sample mean is an estimate of the population mean* 

- What are hypothesis test? How are they useful?

*hypothesis tests allows for us to compare two variables within a dataset and find correlations in the data*

- What is a "null hypothesis?"

*The null hypothesis is the "boring" answer. This serves as the control for the comaprison*

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)